src/ellswift.rs

```diff
@@ -0,0 +1,661 @@
+        mut hash_function: F,
+    ) -> ElligatorSwiftSharedSecret
+    where
+        F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
+    {
+        let mut shared_secret = [0u8; 32];
+        let hashfp = hash_callback::<F>;
+        unsafe {
+            secp256k1_ellswift_xdh(
+                ffi::secp256k1_context_no_precomp,
+                shared_secret.as_mut_c_ptr(),
+                ellswift_a.0.as_c_ptr(),
+                ellswift_b.0.as_c_ptr(),
+                secret_key.as_c_ptr(),
+                party.into(),
+                Some(hashfp),
+                &mut hash_function as *mut F as *mut c_void,
+            );
+        }
+        ElligatorSwiftSharedSecret(shared_secret)
+    }
+    /// Encodes a public key into an ElligatorSwift encoding
+    fn ellswift_encode(pk: PublicKey) -> ElligatorSwift {
+        let mut ell_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        unsafe {
+            secp256k1_ellswift_encode(
+                ffi::secp256k1_context_no_precomp,
+                ell_out.as_mut_c_ptr(),
+                pk.as_c_ptr(),
+                [0u8; 32].as_ptr(),
+            );
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ell_out))
+    }
+    /// Decodes an ElligatorSwift encoding into a [ffi::PublicKey]. This function is a low
+    /// level function, and will be used in internal functions.
+    pub(crate) fn ellswift_decode(ell: ElligatorSwift) -> ffi::PublicKey {
+        unsafe {
+            let mut pk = ffi::PublicKey::new();
+            secp256k1_ellswift_decode(
+                ffi::secp256k1_context_no_precomp,
+                pk.as_mut_c_ptr(),
+                ell.0.as_c_ptr(),
+            );
+            pk
+        }
+    }
+}
+/// The result of [ElligatorSwift::shared_secret], which is a shared secret
+/// computed from the x-only ECDH using both parties' public keys (ElligatorSwift encoded) and our own
+/// private key.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwiftSharedSecret([u8; 32]);
+
+impl Deref for ElligatorSwiftSharedSecret {
+    type Target = [u8; 32];
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+/// Represents which party we are in the ECDH, A is the initiator, B is the responder.
+/// This is important because the hash of the shared secret is different depending on which party
+/// we are. In this context, "we" means the party that is using this library, and possesses the
+/// secret key passed to `ElligatorSwift::shared_secret`.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ElligatorSwiftParty {
+    /// We are the initiator of the ECDH
+    A = 0,
+    /// We are the responder of the ECDH
+    B = 1,
+}
+
+impl From<ElligatorSwiftParty> for i32 {
+    fn from(party: ElligatorSwiftParty) -> Self {
+        match party {
+            ElligatorSwiftParty::A => 0,
+            ElligatorSwiftParty::B => 1,
+        }
+    }
+}
```

Comment:
In 6cde4c32dad779e69cb2242ef46fe0646e22d806:

I'm a bit suspicious of giving these explicit values and having a `i32` conversion. What is the purpose of this? It does not appear to be used in the API.

I guess these are for use with the FFI functions. In this case I'd make an explicit method `to_ffi_int` and not bother assigning values or implementing a `From` impl.